### ISSUE TITLE
Feature Request: Jetbrains (IntelliJ IDEA, Webstorm) Plugin

### ISSUE THREAD

**Issue by siberian-man** [prisma/prisma#1788]
Like the one you have for VS Code but for [IntelliJ](https://www.jetbrains.com/idea/) and/or [WebStorm](https://www.jetbrains.com/webstorm/).

**Comment by nickreynke** [comment]
Aye, thank you!

**Comment by carmenberndt** [comment]
Some features including linting, formatting, jump-to-definition and auto-completion will be done via the Prisma Language Server (related to https://github.com/prisma/vscode/issues/112). Syntax highlighting will require more customized work.

**Comment by Cretezy** [comment]
Would it be possible to ship a interim plugin with the TextMate bundles? I think it would help in the time being, and you can simply change the plugin's implementation when a more permanent one is ready (with the extra features)

**Comment by gabrielcolson** [comment]
I just released the first version of a plugin here: https://github.com/gabrielcolson/intellij-prisma

**Comment by AhmedElywa** [comment]
@gabrielcolson the most needed feature is error report

**Comment by gabrielcolson** [comment]
@AhmedElywa what do you mean by error report? Could you create an issue so we can discuss it?

**Comment by beeman** [comment]
Awesome @gabrielcolson! Yesterday WebStorm notified me about it, that was a great surprise!

**Comment by gabrielcolson** [comment]
Thanks! Do not hesitate to fill an issue if you have a specific feature request 😄

**Comment by olance** [comment]
@gabrielcolson funny how things in the world sometimes happen in the most surprising timings!
Looks like I've started working on a PSL plugin for Jetbrains IDE's the same time you did: our initial commits are literally a day apart! 🙂

So, yeah, there was no Jetbrains plugin until recently, and now there are two 😅 
https://github.com/olance/intellij-prisma

**Comment by italodeandra** [comment]
Still waiting for the official version. It's nice that people want to build their own, but neither of the two existing ones worked for me. And I'm just using the template schema from the documentation. For now I'm using VS just for editing the schemas.

**Comment by olance** [comment]
@italodeandra could you detail what's not working for you?
I've just pushed an update days ago that fixed an issue with maps in enums.
Please feel free to [open an issue](https://github.com/olance/intellij-prisma/issues) for anything missing/buggy, OSS works on feedback too :)

**Comment by italodeandra** [comment]
> @italodeandra could you detail what's not working for you?

@olance Here is an example of what's not working properly (first image). But also the Color Schema looks way better on VSCode (second image).

![image](https://user-images.githubusercontent.com/19225266/109401362-87577d80-794e-11eb-8ed7-f46209abefcc.png)

![image](https://user-images.githubusercontent.com/19225266/109401392-ac4bf080-794e-11eb-9256-5ca05e36146f.png)

**Comment by beeman** [comment]
> > @italodeandra could you detail what's not working for you?
> 
> @olance Here is an example of what's not working properly (first image). But also the Color Schema looks way better on VSCode (second image).
> 
> ![image](https://user-images.githubusercontent.com/19225266/109401362-87577d80-794e-11eb-8ed7-f46209abefcc.png)
> 
> ![image](https://user-images.githubusercontent.com/19225266/109401392-ac4bf080-794e-11eb-9256-5ca05e36146f.png)

The color scheme is not something that comes with the plugin. In my Intellij editors for example it looks way better because of the Intellij Material plug

**Comment by italodeandra** [comment]
> The color scheme is not something that comes with the plugin. In my Intellij editors for example it looks way better because of the Intellij Material plugin

@olance I understand. But the Color Schema isn't a deal breaker to me, as I don't like UI plug-ins. It's just the bug that keeps me not using it.

**Comment by olance** [comment]
Gotcha, I don't know how I missed this syntax (the `MatchedEvent[]`) when writing the grammar!

I'll fix that ASAP :)

I'm also planning on leveling up the error reporting based on prisma's tooling!

**Comment by ivan-ivory** [comment]
@olance Plugin still not working correctly and after the last idea update it stopped working at all.

**Comment by villesau** [comment]
It's a shame that IntelliJ does not have first-class Prisma support. I also opened an issue in Jetbrains issue tracker about it: https://youtrack.jetbrains.com/issue/IDEA-266544

**Comment by olance** [comment]
> @olance Plugin still not working correctly and after the last idea update it stopped working at all.

I know, so sorry about this :(
My job is taking a lot of space right now, I barely have week-ends to work on OSS

**Comment by olance** [comment]
@ivan-ivory @italodeandra I have just pushed update 1.0.3 of my plugin to Jetbrains Plugins, it's awaiting approval!

It just fixes the latests issues for now, hopefully I'll get the time to improve it more throughout the week :)

**Comment by prigara** [comment]
Hello everyone,
I'm from the WebStorm team. I want to thank @olance and @gabrielcolson for creating the plugins.

At this very moment, a native built-in Prisma ORM support is not on our team's roadmap for 2021. Sorry.

We really appreciate the community effort in providing Prisma support in JetBrains IDEs. We'd be glad to answer any of your questions regarding the plugin development for JetBrains IDEs, APIs and best practices. Feel free to get in touch with us: [Slack](https://plugins.jetbrains.com/slack), [my email](mailto:ekaterina.prigara@jetbrains.com).

**Comment by villesau** [comment]
@olance @carmenberndt @janpio I wonder if it would make sense to move https://github.com/olance/intellij-prisma under prisma organization? Maybe Prisma could sponsor Olance to bring the plugin on par with the vscode plugin?

**Comment by janpio** [comment]
That is an interesting idea. The org only has things we fully maintain, and we currently do not have the capacity to do so unfortunately. But helping @olance to spend a bit more time on this, and also having direct people to talk to, would be amazing. And of course that time could also be compensated in some way. Let's talk if you want via jan@prisma.io.

**Comment by olance** [comment]
Hey @villesau & @janpio thanks for this suggestion I'm quite flattered you'd consider this!

I'm interested in seeing how we could make this work, expect an email soon @janpio 🙂

**Comment by sshquack** [comment]
Thanks for creating the community plugins. I heavily use Jetbrains IntelliJ and Webstorm for my everyday dev. Recently noticed that the plugin does not work with the recent IntelliJ releases.

![Screen Shot 2021-08-14 at 11 48 36 AM](https://user-images.githubusercontent.com/5678459/129457395-c72c13d3-815c-48b1-a00c-f7c3b4396456.png)

My workaround is if I have to edit primsa schema, I have to fire up vscode only because the prisma vscode plugin is fantastic. Having a smoother workflow without having to jump IDEs would make my prisma DX much better!

**Comment by Jolg42** [comment]
@sshquack your issue might be resolved now I think, this is the current changelog 
<img width="656" alt="Screen Shot 2021-12-08 at 09 13 35" src="https://user-images.githubusercontent.com/1328733/145172571-e16e12fa-ce2d-4f52-9df5-031c4235c9c0.png">

**Comment by Jolg42** [comment]
As of today I can see 2 extensions on the marketplace

"Prisma"
https://plugins.jetbrains.com/plugin/14240-prisma
https://github.com/gabrielcolson/intellij-prisma

and 

"Prisma Support"
https://plugins.jetbrains.com/plugin/14282-prisma-support
https://github.com/olance/intellij-prisma

Both seem to only have syntax highlighting support for now.

**Comment by villesau** [comment]
@janpio @olance any news? Maybe Jetbrains could chip in to sponsor the plugin too @prigara ?

Currently the missing piece is integration with Prisma language server. I think https://github.com/gtache/intellij-lsp would help with that.

**Comment by olance** [comment]
@villesau thanks for pinging.
@janpio and I have actually exchanged emails a few months back, but I then experienced a (light...) burnout and actually haven't even responded to Jan since then... sorry about that Jan 😕

Thanks for pointing out the language server plugin, I didn't have this in my radar and it could actually simplify so many things! 🙏

The one thing though is that I'm not sure when I'll have the time to get back to this project...

**Comment by villesau** [comment]
Sorry to hear about the burnout! It can happen to anyone and often you realize it only after it's already too late. take care @olance!

Btw, looks like there's also another LSP plugin to choose from: https://github.com/ballerina-platform/lsp4intellij

**Comment by janpio** [comment]
Our documentation now actually links to these 2 plugins that exist:
- https://plugins.jetbrains.com/plugin/14240-prisma
- https://plugins.jetbrains.com/plugin/14282-prisma-support
But I guess they do not give the full feature set yet?

**Comment by AndrewSouthpaw** [comment]
@janpio That's correct. Neither do much beyond syntax highlighting for the `.prisma` file, which is nice to have but pretty minor for me.

Currently the code navigation to a type is broken:

![demo](https://user-images.githubusercontent.com/5187404/157562413-48c53523-8c83-4ebf-9f88-1133795f8692.gif)

If I add the `node_modules/.prisma/client` manually to the `Languages & Frameworks > JavaScript > Libraries`, then it _almost_ gets there but then navigates to a `ModelName` declaration instead of the actual type:

![demo](https://user-images.githubusercontent.com/5187404/157562610-a927b5ff-8e0d-428f-b18a-1ffab5a3d11a.gif)

It's a little perplexing that it lands there, as opposed to the very obvious `User` type at the top of the file...

![image](https://user-images.githubusercontent.com/5187404/157563613-e20bc34b-a44b-40cb-9d98-5c746df72826.png)


Type checking and code completion still work, it's just bothersome to not be able to not be able to hop to a type definition as is possible in most other contexts.

**Comment by janpio** [comment]
Thanks, reports like these are very helpful. The code navigation of Prisma Client usually is not on our radar at all, as in VS Code this all just works out of the box and our VSCode extension is not doing _anything_ to improve on that at all. Seems on Jetbrains products this might be necessary or helpful. (Potentially this can also be fixed by adapting our codes with some annotations or simiar)

**Comment by IGassmann** [comment]
Good news everyone! [WebStorm finally supports Prisma](https://blog.jetbrains.com/webstorm/2022/12/webstorm-2022-3-1/)! [The plugin](https://plugins.jetbrains.com/plugin/20686-prisma-orm) is built by JetBrains

**Comment by Jolg42** [comment]
Indeed! It looks great!

It is available as a plugin: https://plugins.jetbrains.com/plugin/20686-prisma-orm
And the source doe is here: https://github.com/JetBrains/intellij-plugins/tree/master/prisma

Features
- Prisma Schema Language grammar support
- Semantic syntax highlighting
- Code completion
- Documentation
- Line commenter
- Reader mode for comments with Markdown support
- Braces and quotes automatic insertion
- Rename refactoring
- Structure view
- Folding
- LSP support
- Improved completion and resolve for generated Prisma client

**Comment by beeman** [comment]
Awesome that it's officially supported by JetBrains now!

@gabrielcolson thanks a lot for maintaining your plugin, I always enjoyed using it!
